File: src/enigma.jl
# Enigma Machine - Julia Implementation
# Wehrmacht Enigma I (3-rotor, Reflector B, plugboard, double-stepping)
# PeopleTec Inc. - Guinness World Record Attempt 2026

const FWD_W = ["EKMFLGDQVZNTOWYHXUSPAIBRCJ",
               "AJDKSIRUXBLHWTMCQGZNPYFVOE",
               "BDFHJLCPRTXVZNYEIWGAKMUSQO"]
const BWD_W = ["UWYGADFPVZBECKMTHXSLRINQOJ",
               "AJPCZWRLFBDKOTYUQGENHXMIVS",
               "TAGBPCSDQEUFVNZHYIXJWLRKOM"]
const NOTCH_POS = [17, 5, 22]  # 1-based: Q=17, E=5, V=22
const REFL = "YRUHQSLDPXNGOKMIEBFZCWVJAT"

mod26(a) = mod(a + 260, 26)  # always positive

mutable struct Rotor
    fwd::String
    bwd::String
    notch::Int
    offset::Int
end

function make_rotor(num::Int, win::Char)
    Rotor(FWD_W[num], BWD_W[num], NOTCH_POS[num], Int(win) - Int('A') + 1)
end

function fwd_pass(r::Rotor, idx::Int)
    contact = mod26(idx - 1 + r.offset - 1) + 1  # 1-based
    out = Int(r.fwd[contact]) - Int('A') + 1
    mod26(out - 1 - (r.offset - 1)) + 1
end

function bwd_pass(r::Rotor, idx::Int)
    contact = mod26(idx - 1 + r.offset - 1) + 1
    out = Int(r.bwd[contact]) - Int('A') + 1
    mod26(out - 1 - (r.offset - 1)) + 1
end

step!(r::Rotor) = (r.offset = mod(r.offset, 26) + 1)
at_notch(r::Rotor) = r.offset == r.notch

mutable struct Enigma
    left::Rotor
    middle::Rotor
    right::Rotor
    plug::Vector{Int}
end

function make_enigma(rotors, key::String, plugboard::Vector{String}=String[])
    plug = collect(1:26)
    for pair in plugboard
        a = Int(pair[1]) - Int('A') + 1
        b = Int(pair[2]) - Int('A') + 1
        plug[a] = b; plug[b] = a
    end
    Enigma(make_rotor(rotors[1], key[1]),
           make_rotor(rotors[2], key[2]),
           make_rotor(rotors[3], key[3]), plug)
end

function step_rotors!(e::Enigma)
    if at_notch(e.middle)
        step!(e.middle); step!(e.left)
    elseif at_notch(e.right)
        step!(e.middle)
    end
    step!(e.right)
end

function press_key(e::Enigma, c::Char)
    step_rotors!(e)
    idx = Int(c) - Int('A') + 1
    idx = e.plug[idx]
    idx = fwd_pass(e.right, idx)
    idx = fwd_pass(e.middle, idx)
    idx = fwd_pass(e.left, idx)
    idx = Int(REFL[idx]) - Int('A') + 1
    idx = bwd_pass(e.left, idx)
    idx = bwd_pass(e.middle, idx)
    idx = bwd_pass(e.right, idx)
    idx = e.plug[idx]
    Char(idx - 1 + Int('A'))
end

function encrypt(e::Enigma, text::String)
    join([press_key(e, c) for c in uppercase(text) if 'A' <= c <= 'Z'])
end

# Test harness
println("Enigma Machine - Julia Implementation")
println("======================================")

tests = [
    ([1,2,3], "AAA", String[],         "AAAAA",        "BDZGO"),
    ([1,2,3], "AAA", String[],         "HELLOWORLD",   "ILBDAAMTAZ"),
    ([1,2,3], "AAA", String[],         "ATTACKATDAWN", "BZHGNOCRRTCM"),
    ([1,2,3], "MCK", String[],         "HELLOWORLD",   "DLTBBQVPQV"),
    ([3,1,2], "AAA", String[],         "HELLOWORLD",   "KZHDFQYHXT"),
    ([1,2,3], "AAA", ["AB","CD","EF"], "HELLOWORLD",   "IKACBBMTBF"),
]

all_pass = true
for (i, (rotors, key, plugs, plain, expected)) in enumerate(tests)
    e = make_enigma(rotors, key, plugs)
    cipher = encrypt(e, plain)
    ok = cipher == expected
    status = ok ? "PASS" : "FAIL"
    @printf("  Test %d: %-20s -> %-15s [%s]\n", i, plain, cipher, status)
    if !ok
        @printf("          Expected %s, got %s\n", expected, cipher)
        global all_pass = false
    end
end
println(all_pass ? "\n  ALL 6 TESTS PASSED" : "\n  SOME TESTS FAILED")

File: src/enigma.vb
' Enigma Machine - Visual Basic .NET Implementation
' Wehrmacht Enigma I (3-rotor, Reflector B, plugboard, double-stepping)
' PeopleTec Inc. - Guinness World Record Attempt 2026

Module EnigmaModule
    Dim FWD() As String = {"EKMFLGDQVZNTOWYHXUSPAIBRCJ", _
                            "AJDKSIRUXBLHWTMCQGZNPYFVOE", _
                            "BDFHJLCPRTXVZNYEIWGAKMUSQO"}
    Dim BWD() As String = {"UWYGADFPVZBECKMTHXSLRINQOJ", _
                            "AJPCZWRLFBDKOTYUQGENHXMIVS", _
                            "TAGBPCSDQEUFVNZHYIXJWLRKOM"}
    Dim NOTCH_POS() As Integer = {16, 4, 21}
    Dim REFL As String = "YRUHQSLDPXNGOKMIEBFZCWVJAT"

    Function Mod26(a As Integer) As Integer
        Return ((a Mod 26) + 26) Mod 26
    End Function

    Class Rotor
        Public fwd As String, bwd As String
        Public notch As Integer, offset As Integer

        Sub New(num As Integer, win As Char)
            fwd = FWD(num) : bwd = BWD(num)
            notch = NOTCH_POS(num)
            offset = Asc(win) - Asc("A"c)
        End Sub

        Function ForwardPass(idx As Integer) As Integer
            Dim contact As Integer = Mod26(idx + offset)
            Return Mod26(Asc(fwd(contact)) - Asc("A"c) - offset)
        End Function

        Function BackwardPass(idx As Integer) As Integer
            Dim contact As Integer = Mod26(idx + offset)
            Return Mod26(Asc(bwd(contact)) - Asc("A"c) - offset)
        End Function

        Sub StepRotor()
            offset = (offset + 1) Mod 26
        End Sub

        Function AtNotch() As Boolean
            Return offset = notch
        End Function
    End Class

    Class Enigma
        Public left, middle, right As Rotor
        Public plug(25) As Integer

        Sub New(rotors() As Integer, key As String, plugboard() As String)
            left = New Rotor(rotors(0), key(0))
            middle = New Rotor(rotors(1), key(1))
            right = New Rotor(rotors(2), key(2))
            For i As Integer = 0 To 25 : plug(i) = i : Next
            If plugboard IsNot Nothing Then
                For Each pair As String In plugboard
                    Dim a As Integer = Asc(pair(0)) - Asc("A"c)
                    Dim b As Integer = Asc(pair(1)) - Asc("A"c)
                    plug(a) = b : plug(b) = a
                Next
            End If
        End Sub

        Sub StepRotors()
            If middle.AtNotch() Then
                middle.StepRotor() : left.StepRotor()
            ElseIf right.AtNotch() Then
                middle.StepRotor()
            End If
            right.StepRotor()
        End Sub

        Function PressKey(c As Char) As Char
            StepRotors()
            Dim idx As Integer = Asc(c) - Asc("A"c)
            idx = plug(idx)
            idx = right.ForwardPass(idx)
            idx = middle.ForwardPass(idx)
            idx = left.ForwardPass(idx)
            idx = Asc(REFL(idx)) - Asc("A"c)
            idx = left.BackwardPass(idx)
            idx = middle.BackwardPass(idx)
            idx = right.BackwardPass(idx)
            idx = plug(idx)
            Return Chr(idx + Asc("A"c))
        End Function

        Function Encrypt(text As String) As String
            Dim result As String = ""
            For Each c As Char In text.ToUpper()
                If c >= "A"c AndAlso c <= "Z"c Then result &= PressKey(c)
            Next
            Return result
        End Function
    End Class

    Sub Main()
        Console.WriteLine("Enigma Machine - VB.NET Implementation")
        Console.WriteLine("=======================================")

        Dim tests = {
            ({0,1,2}, "AAA", Nothing, "AAAAA", "BDZGO"),
            ({0,1,2}, "AAA", Nothing, "HELLOWORLD", "ILBDAAMTAZ"),
            ({0,1,2}, "AAA", Nothing, "ATTACKATDAWN", "BZHGNOCRRTCM"),
            ({0,1,2}, "MCK", Nothing, "HELLOWORLD", "DLTBBQVPQV"),
            ({2,0,1}, "AAA", Nothing, "HELLOWORLD", "KZHDFQYHXT"),
            ({0,1,2}, "AAA", New String(){"AB","CD","EF"}, "HELLOWORLD", "IKACBBMTBF")
        }

        Dim allPass As Boolean = True
        For i As Integer = 0 To tests.Length - 1
            Dim t = tests(i)
            Dim e As New Enigma(t.Item1, t.Item2, t.Item3)
            Dim cipher As String = e.Encrypt(t.Item4)
            Dim ok As Boolean = cipher = t.Item5
            Dim status As String = If(ok, "PASS", "FAIL")
            Console.WriteLine($"  Test {i+1}: {t.Item4,-20} -> {cipher,-15} [{status}]")
            If Not ok Then Console.WriteLine($"          Expected {t.Item5}") : allPass = False
        Next
        Console.WriteLine(If(allPass, vbLf & "  ALL 6 TESTS PASSED", vbLf & "  SOME TESTS FAILED"))
    End Sub
End Module
